utils/ixdtf/src/parsers/records.rs

```diff
@@ -0,0 +1,201 @@
+///
+/// `IsoParseRecord` is converted into the ISO AST Nodes.
+#[non_exhaustive]
+#[derive(Default, Debug, PartialEq)]
+pub struct IsoParseRecord<'a> {
+    /// Parsed Date Record
+    pub date: Option<DateRecord>,
+    /// Parsed Time
+    pub time: Option<TimeRecord>,
+    /// Parsed UtcOffset
+    pub offset: Option<UTCOffset>,
+    /// Parsed `TimeZone` data (UTCOffset | IANA name)
+    pub tz: Option<TimeZone<'a>>,
+    /// The parsed calendar value.
+    pub calendar: Option<&'a str>,
+    /// Annotations
+    pub annotations: Vec<Annotation<'a>>,
+}
+
+#[non_exhaustive]
+#[derive(Debug, Clone, PartialEq)]
+pub struct Annotation<'a> {
+    pub critical: bool,
+    pub key: &'a str,
+    pub value: &'a str,
+}
+
+#[non_exhaustive]
+#[derive(Default, Debug, Clone, Copy, PartialEq)]
+/// The record of a parsed date.
+pub struct DateRecord {
+    /// Date Year
+    pub year: i32,
+    /// Date Month
+    pub month: u8,
+    /// Date Day
+    pub day: u8,
+}
+
+/// Parsed Time info
+#[non_exhaustive]
+#[derive(Debug, Default, Clone, Copy, PartialEq)]
+pub struct TimeRecord {
+    /// An hour
+    pub hour: u8,
+    /// A minute value
+    pub minute: u8,
+    /// A second value.
+    pub second: u8,
+    /// A millisecond value
+    pub millisecond: u16,
+    /// A Microsecond value
+    pub microsecond: u16,
+    /// A Nanosecond value
+    pub nanosecond: u16,
+}
+
+/// `TimeZone` data
+#[non_exhaustive]
+#[derive(Default, Debug, Clone, PartialEq)]
+pub struct TimeZone<'a> {
+    /// TimeZoneIANAName
+    pub name: Option<&'a str>,
+    /// TimeZoneOffset
+    pub offset: Option<UTCOffset>,
+}
+
+/// A full precision `UtcOffset`
+#[non_exhaustive]
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct UTCOffset {
+    /// The `+`/`-` sign of this `UtcOffset`
+    pub sign: i8,
+    /// The hour value of the `UtcOffset`
+    pub hour: u8,
+    /// The minute value of the `UtcOffset`.
+    pub minute: u8,
+    /// The second value of the `UtcOffset`.
+    pub second: u8,
+    /// Any millisecond value of the `UTCOffset`
+    pub millisecond: u16,
+    /// Any microsecond value of the `UTCOffset`
+    pub microsecond: u16,
+    /// Any nanosecond value of the `UTCOffset`
+    pub nanosecond: u16,
+}
+
+/// The resulting record of a `Duration` parse.
+#[non_exhaustive]
+#[cfg(feature = "duration")]
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub struct DurationParseRecord {
+    /// Duration Sign
+    pub years: i32,
+    /// A `DateDuration` record.
+    pub months: i32,
+    /// The `weeks` value.
+    pub weeks: i32,
+    /// The `days` value.
+    pub days: i32,
+    /// The `hours` value.
+    pub hours: f64,
+    /// The `minutes` value.
+    pub minutes: f64,
+    /// The `seconds` value.
+    pub seconds: f64,
+    /// The `milliseconds` value.
+    pub milliseconds: f64,
+    /// The `microseconds` value.
+    pub microseconds: f64,
+    /// The `nanoseconds` value.
+    pub nanoseconds: f64,
```

Comment:
Discussionm, here and below: it looks like you made these public fields `f64` since they could be either integral or fractional in the ISO Duration syntax. Is that right? Are there cases where the `f64` could lose precision? Did you consider other ways of expressing these quantities?

Reply:
I had partially played around with changing these fields. They were mostly the drivers for keeping the fraction parsing as allocating to a `String` that was parsed into a `f64` over a `u32` approach (P1.005H is different than 10:12:08.000123456). I'll take a look at this again with the fraction change / any refactor.

I'm not entirely sure if there are cases where `f64` loses precision. The fraction part can only be 9 decimal places, at least going according the grammar laid out in Temporal. No matter how much research I did, I wasn't able to find any upper and lower bounds on the duration field values (outside of the restriction on Duration itself).

If you have any ideas and/or suggestions, I'd definitely be open to them.

Reply:
I think there may be an option to simply parse the components and provide them as one record without manually handling the fraction part, which would simplify everything.

```rust

pub struct Duration {
  pub years: i32,
  pub months: i32,
  pub weeks: i32,
  pub days: i32,
  pub hours: Option<i32>,
  pub minutes: Option<i32>,
  pub seconds: Option<i32>,
  pub fraction: Option<(u32, u8)>
}
```

The above would essentially be true where the FractionPart applies to the smallest available unit.

Reply:
Yeah I like something along those lines, or maybe instead of having all the `Option`s, use an enum

```rust
pub enum DurationFraction {
    Hours(u32),
    Minutes(u32),
    Seconds(u32),
}
```